packages/flutter/lib/src/widgets/widget_state.dart

```diff
@@ -594,6 +749,96 @@ class _WidgetStatePropertyWith<T> implements WidgetStateProperty<T> {
   T resolve(Set<WidgetState> states) => _resolve(states);
 }
 
+/// A [Map] used to resolve to a single value of type `T` based on
+/// the current set of Widget states.
+///
+/// {@template flutter.widgets.WidgetStateProperty.WidgetStateMap}
+/// Example:
+///
+/// ```dart
+/// WidgetStateProperty<Color?>.map(<WidgetStateMapKey, Color?>{
+///   WidgetState.error: Colors.red,
+///   WidgetState.hovered & WidgetState.focused: Colors.blueAccent,
+///   WidgetState.focused: Colors.blue,
+///   ~WidgetState.disabled: Colors.black,
+/// });
+///
+/// // the same implementation, but with a WidgetPropertyResolver:
+/// WidgetStateProperty.resolveWith<Color?>((Set<WidgetState> states) {
+///   if (states.contains(WidgetState.error)) {
+///     return Colors.red;
+///   } else if (states.contains(WidgetState.hovered) && states.contains(WidgetState.focused)) {
+///     return Colors.blueAccent;
+///   } else if (states.contains(WidgetState.focused)) {
+///     return Colors.blue;
+///   } else if (!states.contains(WidgetState.disabled)) {
+///     return Colors.black;
+///   }
+///   return null;
+/// });
+/// ```
+///
+/// A widget state combination can be stored in a variable,
+/// and [WidgetState.any] can be used for non-nullable types:
+///
+/// ```dart
+/// final WidgetStateMapKey selectedError = WidgetState.selected & WidgetState.error;
+///
+/// final WidgetStateProperty<Color> color = WidgetStateProperty<Color>.map(
+///   <WidgetStateMapKey, Color>{
+///     selectedError & WidgetState.hovered: Colors.redAccent,
+///     selectedError: Colors.red,
+///     WidgetState.any: Colors.black,
+///   },
+/// );
+///
+/// // WidgetPropertyResolver implementation:
+/// final WidgetStateProperty<Color> colorResolveWith = WidgetStateProperty.resolveWith<Color>(
+///   (Set<WidgetState> states) {
+///     if (states.containsAll(<WidgetState>{WidgetState.selected, WidgetState.error})) {
+///       if (states.contains(WidgetState.hovered)) {
+///         return Colors.redAccent;
+///       }
+///       return Colors.red;
+///     }
+///     return Colors.black;
+///   },
+/// );
+/// ```
+/// {@endtemplate}
+typedef WidgetStateMap<T> = Map<WidgetStateMapKey, T>;
+
+// A private class, used to create the [WidgetStateProperty.map] constructor.
+class _WidgetStateMapper<T> implements WidgetStateProperty<T> {
+  const _WidgetStateMapper(this.map);
+
+  final WidgetStateMap<T> map;
+
+  @override
+  T resolve(Set<WidgetState> states) {
+    for (final MapEntry<WidgetStateMapKey, T> entry in map.entries) {
```

Comment:
I am missing documentation stating in which order it will try to match the entries of the provided map. Can you add this to some prominent places?

Reply:
Good point. Since this is a private class's method, I'll document the behavior in each constructor that uses it.